### PR TITLE
New bootstrap

### PR DESCRIPTION
**Pull request by jay-aye-see-kay**
To run this locally (this is different from what goes in the readme, that doesn't have a branch name so won't work until this PR merges):
```sh
sh -c "$(curl -fsSL "https://raw.githubusercontent.com/cultureamp/devbox-extras/new-bootstrap/scripts/install_hotel.sh")"
```

Uninstall script has been useful for local testing/dev. But not sure if we should merge it. I'm leaning towards no because it won't help somebody remove their LDE setup, which is potentially what someone might try if their setup was borked. It won't uninstall nix or devbox or anything that hotel installs and maintains - only hotel itself. But also if it's not documented anywhere it's unlikely to even get run by mistake, and I don't think can cause any harm

### CODE REVIEW

**Comment by t-monaghan** [comment]
My first run on a clean VM gave:
`mv rename hotel to /Users/username/.local/share/hotel/bin/hotel: No such file or directory`
I suspect this requires the `/Users/username/.local/share/hotel/bin` directory to exist. I'll have a shot at a fix now.

File: scripts/install_hotel.sh [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+install_hotel() {
+  log "=> installing hotel, this will ask for a sudo password"
+  INITIAL_DIR="$PWD"
+  TMPDIR=$(mktemp -d)
+  cd "$TMPDIR"
+  download_latest_hotel "$1"
+  mkdir -p "$hotel_bin_path"
+  mv hotel "$hotel_bin_path"
+  cd "$INITIAL_DIR"
+  rm -rf "$TMPDIR"
```
Comment (t-monaghan) [inline scripts/install_hotel.sh:137]:
```suggestion
  sudo mv hotel "$hotel_bin_path"
  cd "$INITIAL_DIR"
  rm -rf "$TMPDIR"
  sudo mkdir /usr/local/bin
```
The above suggestion is a possible solution to get the installer to run on the first try. 
In a fresh VM with macos I found there was no `/usr/local/bin` dir. However I'm confused as to why the `mv` command requires `sudo`.

File: scripts/install_hotel.sh [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+  if [ "$scopes" = "" ]; then
+    log "token not valid"
+    return 1
+  fi
+  if ! echo "$scopes" | grep "repo" >/dev/null 2>&1; then
+    log "token does not have the 'repo' scope"
+    return 1
+  fi
+}
+# }}}
```
Comment (t-monaghan) [inline scripts/install_hotel.sh:69]: Is this a leftover?

**Review by t-monaghan — CHANGES_REQUESTED** [review]
Nice work!
Is it worth setting up an e2e test on this script like we have in hotel?

File: scripts/install_hotel.sh (hunk 1 above)
Comment (jay-aye-see-kay) [reply scripts/install_hotel.sh:137]: Good catch, have fixed inside hotel, can you review this one https://github.com/cultureamp/hotel/pull/39

**Comment by jay-aye-see-kay** [comment]
> My first run on a clean VM had an error after providing the gh token: `mv rename hotel to /Users/username/.local/share/hotel/bin/hotel: No such file or directory` I suspect this requires the `/Users/username/.local/share/hotel/bin` directory to exist. I'll have a shot at a fix now.

I've replicated this issue in the github action - for me it was caused by a github token without the correct permissions, which results in no `hotel` binary to move to the new location. I'll improve that error message

**Comment by jay-aye-see-kay** [comment]
e2e tests brought up some interesting edge cases
- fine-grained access tokens don't have any way to query their scopes, so got rid of the check to see if it was approximately valid
- now telling curl to fail when getting list of releases and checking the exit code - this gives much less misleading errors when the token isn't valid
- removed the option to re-use an existing stored token, this is a rare use case, more useful for dev than users, and was causing issues in the tests - better to keep the tests simple

**Comment by t-monaghan** [comment]
The flow is broken in VM testing. After providing what I believe to be a valid token I get the below output. The success message printed suspiciously quickly after providing the password. Hotel is not on path.

I'm not sure what's gone wrong here. I've been looking through [this diff to find the culprit](https://github.com/cultureamp/devbox-extras/pull/26/files/3f694aad7ebc15acf473c16893ec0fa6a861135e..75dfe8fa850f261d1636d46dddc8a84238e44fb9?diff=split&w=0).

image

**Comment by jay-aye-see-kay** [comment]
Tried it in a VM too - for some reason setting `/usr/local/bin` as `644` makes `hotel` un-executable. See the following screenshot where after installing it I set that dir as `755` and `hotel` is working again (the fix is second to last line)

Screenshot 2024-06-06 at 9 42 15 PM

This won't ever affect a CA laptop as they all come pre-installed with `jamf` which is in that directory, so `hotel` won't ever create it -> it's safe to merge and use this as is.

Here's the fix https://github.com/cultureamp/hotel/pull/42

**Review by t-monaghan — APPROVED** [review]
Working great on the VM now. Thanks for pushing the fix.
